pag-lexer/src/normalization.rs

```diff
@@ -6,182 +6,193 @@
-                return r;
-            }
-            if matches!(&*r, Top) {
-                return s;
-            }
+    };
+}
 
-            let ordering = r.cmp(&s);
-            if ordering.is_eq() {
-                return r;
-            }
-            if let (Set(r), Set(s)) = (&*r, &*s) {
-                return match r.intersection(s) {
-                    Some(set) => Rc::new(Set(set)),
-                    None => Rc::new(Bottom),
-                };
-            }
-            // always right heavy
-            if let Intersection(r1, r2) = &*r {
-                return normalize(Rc::new(Intersection(
-                    r1.clone(),
-                    Rc::new(Intersection(r2.clone(), s)),
-                )));
-            }
+recursive_flatten! {
+    flatten_concat, Concat
+}
 
-            // will not fall in any of the above cases -- it is safe to return
-            if ordering.is_lt() || matches!(&*s, Intersection(..)) {
-                Rc::new(Intersection(r, s))
-            } else {
-                Rc::new(Intersection(s, r))
-            }
-        }
-        Set(x) => {
-            if x.is_full_set() {
-                Rc::new(Top)
-            } else {
-                tree
-            }
-        }
-        Concat(r, s) => {
-            let nr = normalize(r.clone());
-            let ns = normalize(s.clone());
-            // empty . s = empty
-            if matches!(&*nr, Bottom) {
-                return nr;
-            }
-            // r . empty = empty
-            if matches!(&*ns, Bottom) {
-                return ns;
-            }
-            // r. epsilon = r
-            if matches!(&*ns, Epsilon) {
-                return nr;
-            }
-            // epsilon . s = s
-            if matches!(&*nr, Epsilon) {
-                return ns;
-            }
-            if let Concat(r1, r2) = &*nr {
-                return normalize(Rc::new(Concat(r1.clone(), Rc::new(Concat(r2.clone(), ns)))));
-            }
+recursive_flatten! {
+    flatten_union, Union
+}
 
-            if Rc::ptr_eq(&nr, r) && Rc::ptr_eq(&ns, s) {
-                return tree;
+recursive_flatten! {
+    flatten_intersection, Intersection
+}
+
+fn sequence_unchanged(x: &[Rc<RegexTree>], y: &[Rc<RegexTree>]) -> bool {
+    x.iter()
+        .map(|x| x.as_ref() as *const RegexTree)
+        .eq(y.iter().map(|y| y.as_ref() as *const RegexTree))
+}
+
+pub fn normalize(node: Rc<RegexTree>) -> Rc<RegexTree> {
+    match node.as_ref() {
+        Bottom | Epsilon | Set(..) => node,
+        Concat(old) => {
+            let new: RcVec = old.iter().map(|x| normalize(x.clone())).collect();
+            let new: RcVec = flatten_concat(&new);
+            if new.iter().any(|x| matches!(x.as_ref(), Bottom)) {
+                return RegexTree::bottom();
+            }
+            let new: RcVec = new
+                .into_iter()
+                .filter(|x| !matches!(x.as_ref(), Epsilon))
+                .collect();
```

Comment:
I believe we can combine the first two `let new` into one with no extra clones. And if you don't hurry to create Rc, you can save the third `let new` and the `new[0].clone()` below.